llvm/lib/Target/SPIRV/SPIRVEmitIntrinsics.cpp

```diff
@@ -748,7 +748,16 @@ Type *SPIRVEmitIntrinsics::deduceElementTypeHelper(
     if (II && II->getIntrinsicID() == Intrinsic::spv_resource_getpointer) {
       auto *ImageType = cast<TargetExtType>(II->getOperand(0)->getType());
       assert(ImageType->getTargetExtName() == "spirv.Image");
-      Ty = ImageType->getTypeParameter(0);
+      for (auto U : II->users()) {
+        if (auto *LD = dyn_cast<LoadInst>(U)) {
+          Ty = LD->getType();
+        } else if (auto *ST = dyn_cast<StoreInst>(U)) {
+          Ty = ST->getAccessType();
+        } else {
+          llvm_unreachable("Unexpected user. The only expect users of a "
+                           "resource pointer to an image are loads and stores");
+        }
+      }
```

Comment:
Aren't we expected to have a single user per getpointer instruction?
If so, I'd be in favor of an assert on `hasOneUser()`, and then explicitly load the first user.

Reply:
I figured if there were different, then we would get an error because of a bitcast later, but a closer assert is better.

Reply:
I updated the code. I removed the loop, and look at just the first use. I also use `getAccessType` from the instruction class. This should cover more cases that can come up (atomic operations on RWBuffers). However, there are no tests yet because that has not been implemented in HLSL. I don't know exactly what it will look like.